src/drivers/imx/dummy-dma.c

```diff
@@ -3,28 +3,116 @@
+	tracev_event(TRACE_CLASS_DMA, __e, ##__VA_ARGS__)
+#define trace_dummydma_error(__e, ...) \
+	trace_error(TRACE_CLASS_DMA, __e, ##__VA_ARGS__)
+
+/* Dummy DMA driver (software-based DMA controller)
+ *
+ * This driver is usable on all platforms where the DSP can directly access
+ * all of the host physical memory (or at least the host buffers).
+ *
+ * The way this driver works is that it simply performs the copies
+ * synchronously within the dma_start() and dma_copy() calls.
+ *
+ * One of the drawbacks of this driver is that it doesn't actually have a true
+ * IRQ context, as the copy is done synchronously and the IRQ callbacks are
+ * called in process context.
+ *
+ * An actual hardware DMA driver may be preferable because of the above
+ * drawback which comes from a software implementation. But if there isn't any
+ * hardware DMA controller dedicated for the host this driver can be used.
+ *
+ * This driver requires physical addresses in the elems. This assumption only
+ * holds if you have CONFIG_HOST_PTABLE enabled, at least currently.
+ */
+
+struct dma_chan_pdata {
+	struct dma_sg_elem_array *elems;
+	int sg_elem_curr_idx;
+	uintptr_t r_pos, w_pos;
+};
+
+/* allocate next free DMA channel; we will ignore req_chan because
+ * channel numbers are not meaningful
+ */
 static struct dma_chan_data *dummy_dma_channel_get(struct dma *dma,
 						   unsigned int req_chan)
 {
+	struct dma_chan_data *chans = dma_get_drvdata(dma);
+	uint32_t flags;
+	int channel;
+
+	spin_lock_irq(dma->lock, flags);
+	for (channel = 0; channel < dma->plat_data.channels; channel++) {
+		/* use channel if it's free */
+		if (chans[channel].status == COMP_STATE_INIT) {
+			chans[channel].status = COMP_STATE_READY;
+
+			atomic_add(&dma->num_channels_busy, 1);
+
+			/* return channel */
+			spin_unlock_irq(dma->lock, flags);
+			return &chans[channel];
+		}
+	}
+	spin_unlock_irq(dma->lock, flags);
+	trace_dummydma_error("dummy-dmac: %d no free channel",
+			     dma->plat_data.id);
 	return NULL;
 }
 
+static void dummy_dma_channel_put_unlocked(struct dma_chan_data *channel)
+{
+	struct dma_chan_pdata *ch = dma_chan_get_data(channel);
+
+	/* Reset channel state */
+	channel->cb = NULL;
+	channel->cb_type = 0;
+	channel->cb_data = NULL;
+
+	ch->elems = NULL;
+	channel->desc_count = 0;
+	ch->sg_elem_curr_idx = 0;
+
+	ch->r_pos = 0;
+	ch->w_pos = 0;
+
+	channel->status = COMP_STATE_INIT;
+	atomic_sub(&channel->dma->num_channels_busy, 1);
+}
+
 /* channel must not be running when this is called */
 static void dummy_dma_channel_put(struct dma_chan_data *channel)
 {
+	uint32_t flags;
+
+	spin_lock_irq(channel->dma->lock, flags);
+	dummy_dma_channel_put_unlocked(channel);
+	spin_unlock_irq(channel->dma->lock, flags);
 }
 
+static int docopies(struct dma_chan_data *ch, int bytes);
+
 static int dummy_dma_start(struct dma_chan_data *channel)
 {
-	return 0;
+	/* TODO should we actually do any copying in start()? */
+	return docopies(channel, 0);
 }
```

Comment:
I think start should also be a no-op, right?

Reply:
Start cannot be a no-op, as that will fail preload and the transfer of pages. Will remove the TODO and add another comment stating this.